### PR TITLE
Centralize the cache ID in CacheConfig

### PR DESCRIPTION
**Pull request by ekinanp**
This way, CachedList, CachedOpen, and CachedMetadata do not require the
caller to pass-in a cache ID. Thus, plugin authors are free to use them.
CachedMetadata is especially useful for a container/ec2 instance's metadata.json
file.

This commit also paves the way for creating a more general CachedOp
method that plugin authors can use if they need more fine-grained
caching.

### CODE REVIEW

File: plugin/cache.go [hunk 1]
```diff
@@ -134,7 +149,11 @@ func ClearCacheFor(path string) ([]string, error) {
 	return cache.Delete(rx), nil
 }
 
-func cachedOpHelper(op cachedOp, entry Entry, id string, generateValue func() (interface{}, error)) (interface{}, error) {
+func panicNilCacheConfig() {
```
Comment (ekinanp) [inline plugin/cache.go]: I wonder if we can just embed `CacheConfig` in the `Entry` interface (i.e. have its methods be part of the `Entry` interface). We can force plugin authors to use EntryBase everywhere (and use `plugin.NewEntry` to create these objects). Embedding `CacheConfig` in `Entry` would avoid the nil cache config case, and it makes it easy to reason about stuff (all entries are cacheable and have cache IDs, but you can choose to turn off caching if you don't need the behavior)

**Comment by ekinanp** [comment]
Exposing `CachedOp` would also solve https://github.com/puppetlabs/wash/blob/master/volume/dir.go#L13 -- you'd be able to create a `CachedDirMap` method and pass-around the main volume object across `VolumeDir` and `VolumeFile`

File: plugin/cache.go (hunk 1 above)
Comment (ekinanp) [reply plugin/cache.go]: Actually, we could just have `GetTTLOf`, and `ID` added to Entry while the setters could be restricted to EntryBase.

I don't see a good reason why plugin authors shouldn't use EntryBase all the time.

File: plugin/types.go [hunk 2]
```diff
@@ -125,8 +120,43 @@ type Attributes struct {
 // SizeUnknown can be used to denote that the size is unknown and should be queried from content.
 const SizeUnknown = ^uint64(0)
 
-// The Registry contains the core filesystem data.
-// Plugins: maps plugin mount points to their implementations.
+// Registry represents the plugin registry. It is also Wash's root.
```
Comment (MikaelSmith) [inline plugin/types.go]: I'd put this n a separate file now. It's something plugin writers don't need to know, and I've been trying to keep types.go constrained to things you use when writing plugins.

File: plugin/cache.go (hunk 1 above)
Comment (MikaelSmith) [reply plugin/cache.go]: That seems ok.

Comment (ekinanp) [reply plugin/cache.go]: I'll do that in this PR as a separate commit

**Pull request merged**